components/match2/wikis/callofduty/match_group_input_custom.lua

```diff
@@ -0,0 +1,499 @@
+			end
+		end
+	end
+
+	-- see if match should actually be finished if bestof limit was reached
+	if isScoreSet and not Logic.readBool(match.finished) then
+		local firstTo = math.ceil(match.bestof / 2)
+		for _, item in pairs(opponents) do
+			if (tonumber(item.score or 0) or 0) >= firstTo then
+				match.finished = true
+				break
+			end
+		end
+	end
+
+	-- check if match should actually be finished due to a non score status
+	if not Logic.readBool(match.finished) then
+		for _, opponent in pairs(opponents) do
+			if String.isNotEmpty(opponent.status) and opponent.status ~= 'S' then
+				match.finished = true
+				break
+			end
+		end
+	end
+
+	-- see if match should actually be finished if score is set
+	if isScoreSet and not Logic.readBool(match.finished) and match.hasDate then
+		local currentUnixTime = os.time(os.date('!*t'))
+		local lang = mw.getContentLanguage()
+		local matchUnixTime = tonumber(lang:formatDate('U', match.date))
+		local threshold = match.dateexact and SECONDS_UNTIL_FINISHED_EXACT
+			or SECONDS_UNTIL_FINISHED_NOT_EXACT
+		if matchUnixTime + threshold < currentUnixTime then
+			match.finished = true
+		end
+	end
+
+	-- apply placements and winner if finshed
+	if not String.isEmpty(match.winner) or Logic.readBool(match.finished) then
+		match, opponents = CustomMatchGroupInput.getResultTypeAndWinner(match, opponents)
+	end
+
+	-- Update all opponents with new values
+	for opponentIndex, opponent in pairs(opponents) do
+		match['opponent' .. opponentIndex] = opponent
+	end
+	return match
+end
+
+-- Get Playerdata from Vars (get's set in TeamCards) for team opponents
+function matchFunctions.getTeamPlayers(match, opponentIndex, teamName)
+	-- match._storePlayers will break after the first empty player. let's make sure we don't leave any gaps.
+	local count = 1
+	for playerIndex = 1, MAX_NUM_PLAYERS do
+		-- parse player
+		local player = Json.parseIfString(match['opponent' .. opponentIndex .. '_p' .. playerIndex]) or {}
+		player.name = player.name or Variables.varDefault(teamName .. '_p' .. playerIndex)
+		player.flag = player.flag or Variables.varDefault(teamName .. '_p' .. playerIndex .. 'flag')
+		player.displayname = player.displayname or Variables.varDefault(teamName .. '_p' .. playerIndex .. 'dn')
+		if not Table.isEmpty(player) then
+			match['opponent' .. opponentIndex .. '_p' .. count] = player
+			count = count + 1
+		end
+	end
+	return match
+end
+
+-- Get Playerdata for non-team opponents
+function matchFunctions.getPlayers(match, opponentType, opponentIndex)
+	local players = {}
+	for playerIndex = 1, MAX_NUM_PLAYERS do
+		-- parse player
+		local player = Json.parseIfString(match['opponent' .. opponentIndex .. '_p' .. playerIndex]) or {}
+		player.name = player.name or 'TBD'
+		player.flag = player.flag
+		player.displayname = player.displayname or player.name
+		if Table.isNotEmpty(player) then
+			table.insert(players, player)
+		end
+	end
+
+	return players
+end
```

Comment:
this just breaks solo opponent entirely
since this function is only used for solo opponents you can remove the loop entirely as well as clean it up
additionally you can rename it to be more fitting for its usage

```suggestion
-- Get Playerdata for solo opponents
function matchFunctions.readSoloOpponent(match, opponentIndex)
	local player = Json.parseIfString(match['opponent' .. opponentIndex .. '_p' .. playerIndex]) or {}
	player.name = player.name or 'TBD'
	player.flag = player.flag
	player.displayname = player.displayname or player.name

	return {player}
end
```

Reply:
would the playerIndex on 427 in the current revision need rename? (or a variable to be called on the top of the module?)

Reply:
see comments below